packages/uif/src/indexers/multi/MultiIndexer.ts

```diff
@@ -0,0 +1,208 @@
+   * configurations are persisted.
+   */
+  abstract saveConfigurations(
+    configurations: SavedConfiguration<T>[],
+  ): Promise<void>
+
+  async initialize(): Promise<number> {
+    const saved = await this.multiInitialize()
+    const { toRemove, toSave, safeHeight } = diffConfigurations(
+      this.configurations,
+      saved,
+    )
+    this.saved = toSave
+    if (toRemove.length > 0) {
+      await this.removeData(toRemove)
+    }
+    await this.saveConfigurations(toSave)
+    return safeHeight
+  }
+
+  async update(from: number, to: number): Promise<number> {
+    const range = findRange(this.ranges, from)
+    if (range.configurations.length === 0) {
+      return Math.min(range.to, to)
+    }
+
+    const { configurations, minCurrentHeight } = getConfigurationsInRange(
+      range,
+      this.saved,
+      from,
+    )
+    const adjustedTo = Math.min(range.to, to, minCurrentHeight)
+
+    this.logger.info('Calling multiUpdate', {
+      from,
+      to: adjustedTo,
+      configurations: configurations.length,
+    })
+    const newHeight = await this.multiUpdate(from, adjustedTo, configurations)
+    if (newHeight < from || newHeight > adjustedTo) {
+      throw new Error(
+        'Programmer error, returned height must be between from and to (both inclusive).',
+      )
+    }
+
+    if (newHeight > from) {
+      updateSavedConfigurations(this.saved, configurations, newHeight)
+      await this.saveConfigurations(this.saved)
+    }
+
+    return newHeight
+  }
+
+  async invalidate(targetHeight: number): Promise<number> {
+    return Promise.resolve(targetHeight)
+  }
+
+  async setSafeHeight(): Promise<void> {
+    return Promise.resolve()
+  }
+}
+
+function findRange<T>(
+  ranges: ConfigurationRange<T>[],
+  from: number,
+): ConfigurationRange<T> {
+  const range = ranges.find((range) => range.from <= from && range.to >= from)
+  if (!range) {
+    throw new Error('Programmer error, there should always be a range')
+  }
+  return range
+}
+
+function getConfigurationsInRange<T>(
+  range: ConfigurationRange<T>,
+  savedConfigurations: SavedConfiguration<T>[],
+  currentHeight: number,
+): { configurations: UpdateConfiguration<T>[]; minCurrentHeight: number } {
+  let minCurrentHeight = Infinity
+  const configurations = range.configurations.map(
+    (configuration): UpdateConfiguration<T> => {
+      const saved = savedConfigurations.find((c) => c.id === configuration.id)
+      if (
+        // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
+        saved &&
+        saved.currentHeight !== null &&
+        saved.currentHeight > currentHeight
+      ) {
+        minCurrentHeight = Math.min(minCurrentHeight, saved.currentHeight)
+        return { ...configuration, hasData: true }
+      } else {
+        return { ...configuration, hasData: false }
+      }
+    },
+  )
+  return { configurations, minCurrentHeight }
+}
+
+function updateSavedConfigurations<T>(
```

Comment:
Does this function respect not updating configs that are already synced to the top (until = last synced)?